cloud/scope/powervs_cluster.go

```diff
@@ -871,6 +873,329 @@ func (s *PowerVSClusterScope) createDHCPServer() (*string, error) {
+			}
+			s.Logger.Info("Creating security group rule", "securityGroupID", *securityGroupID, "direction", *direction, "protocol", *protocol, "securityGroup", *remote.SecurityGroupName, "securityGroupCRN", *sg.CRN)
+			remoteOption.CRN = sg.CRN
+		default:
+			s.Logger.Info("Creating security group rule", "securityGroupID", *securityGroupID, "direction", *direction, "protocol", *protocol, "cidr", "0.0.0.0/0")
+			remoteOption.CIDRBlock = ptr.To("0.0.0.0/0")
+		}
+
+		return nil
+	}
+
+	remoteOption := &vpcv1.SecurityGroupRuleRemotePrototype{}
+	if err := setRemote(remote, remoteOption); err != nil {
+		return fmt.Errorf("error while setting remote option to create security group rule: %v", err)
+	}
+
+	options := vpcv1.CreateSecurityGroupRuleOptions{
+		SecurityGroupID: securityGroupID,
+	}
+
+	options.SetSecurityGroupRulePrototype(&vpcv1.SecurityGroupRulePrototype{
+		Direction: direction,
+		Protocol:  protocol,
+		PortMin:   portMin,
+		PortMax:   portMax,
+		Remote:    remoteOption,
+	})
+
+	if _, _, err := s.IBMVPCClient.CreateSecurityGroupRule(&options); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// createVPCSecurityGroupRules creates rules for a security group.
+func (s *PowerVSClusterScope) createVPCSecurityGroupRules(securityGroupSpec infrav1beta2.SecurityGroup, securityGroupID *string) error {
+	for _, rule := range securityGroupSpec.Rules {
+		var protocol *string
+		var portMax, portMin *int64
+
+		direction := ptr.To(string(rule.Direction))
+		switch rule.Direction {
+		case infrav1beta2.SecurityGroupRuleDirectionInbound:
+			protocol = ptr.To(string(rule.Source.Protocol))
+			if rule.Source.PortRange != nil {
+				portMin = ptr.To(int64(rule.Source.PortRange.MinimumPort))
+				portMax = ptr.To(int64(rule.Source.PortRange.MaximumPort))
+			}
+
+			for _, remote := range rule.Source.Remotes {
+				if err := s.createVPCSecurityGroupRule(securityGroupID, direction, protocol, portMin, portMax, remote); err != nil {
+					return fmt.Errorf("error creating security group rule: %v", err)
+				}
+			}
+		case infrav1beta2.SecurityGroupRuleDirectionOutbound:
+			protocol = ptr.To(string(rule.Destination.Protocol))
+			if rule.Destination.PortRange != nil {
+				portMin = ptr.To(int64(rule.Destination.PortRange.MinimumPort))
+				portMax = ptr.To(int64(rule.Destination.PortRange.MaximumPort))
+			}
+
+			for _, remote := range rule.Destination.Remotes {
+				if err := s.createVPCSecurityGroupRule(securityGroupID, direction, protocol, portMin, portMax, remote); err != nil {
+					return fmt.Errorf("error creating security group rule: %v", err)
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+// createVPCSecurityGroup creates a VPC security group.
+func (s *PowerVSClusterScope) createVPCSecurityGroup(spec infrav1beta2.SecurityGroup) (*string, error) {
+	s.Logger.Info("Creating security group", "name", *spec.Name)
+
+	options := &vpcv1.CreateSecurityGroupOptions{
+		VPC: &vpcv1.VPCIdentity{
+			ID: s.GetVPCID(),
```

Comment:
As per the reconcile order, when it reaches here VPC status would have been populated right?
Do we really need to check? If its not populated, it would requeue on ReconcileVPC only I believe.